rm/rm_ha.c

```diff
@@ -416,8 +416,12 @@ M0_INTERNAL void m0_rm_ha_unsubscribe(struct m0_rm_ha_tracker *tracker)
 {
 	struct m0_conf_obj *obj;
 
-	if (m0_clink_is_armed(&tracker->rht_conf_exp))
-		m0_clink_cleanup_locked(&tracker->rht_conf_exp);
+	if (m0_clink_is_armed(&tracker->rht_conf_exp)) {
+		if (m0_chan_is_locked(tracker->rht_conf_exp.cl_chan))
+			m0_clink_cleanup_locked(&tracker->rht_conf_exp);
+		else
+			m0_clink_cleanup(&tracker->rht_conf_exp);
+	}
```

Comment:
A situation with data structure can be either locked or unlocked is unusual. Please document why and when is this possible.

Reply:
Without this it panics on running `rm-ut`:

```
$ ./scripts/m0 run-ut -t rm-ut
----- run_ut -t rm-ut -----
rm-ut 
  api                                             0.00 sec    4 KiB
  lcredits                                        0.00 sec   18 KiB
  fop-funcs                                       0.00 sec   35 KiB
  fom-funcs                                       0.32 sec   82 MiB
  rmsvc                                           1.28 sec  179 MiB
  flock  motr[02222]:  3510  FATAL  [lib/assert.c:50:m0_panic]  panic: (m0_chan_is_locked(link->cl_chan)) at m0_clink_cleanup_locked() (lib/chan.c:321)  [git: sage-base-1.0-172-gc0812e366] /var/motr/m0ut/m0trace.2222
Motr panic: (m0_chan_is_locked(link->cl_chan)) at m0_clink_cleanup_locked() lib/chan.c:321 (errno: 0) (last failed: none) [git: sage-base-1.0-172-gc0812e366] pid: 2222  /var/motr/m0ut/m0trace.2222
...
$ sudo gdb /var/crash/core.0.lt-m0ut.2222 
...
(gdb) bt
#0  0x00007f8ac8b10387 in raise () from /lib64/libc.so.6
#1  0x00007f8ac8b11a78 in abort () from /lib64/libc.so.6
#2  0x00007f8aca367e25 in m0_arch_panic (c=c@entry=0x7f8aca7efc00 <__pctx.5769>, ap=ap@entry=0x7f8a193f3588)
    at lib/user_space/uassert.c:131
#3  0x00007f8aca356ee4 in m0_panic (ctx=ctx@entry=0x7f8aca7efc00 <__pctx.5769>) at lib/assert.c:52
#4  0x00007f8aca3586ea in m0_clink_cleanup_locked (link=link@entry=0x7f8ac00f5780) at lib/chan.c:321
#5  0x00007f8aca3d6350 in m0_rm_ha_unsubscribe (tracker=tracker@entry=0x7f8ac00f5728) at rm/rm_ha.c:420
#6  0x00007f8aca3d6382 in m0_rm_ha_unsubscribe_lock (tracker=tracker@entry=0x7f8ac00f5728) at rm/rm_ha.c:441
#7  0x00007f8aca3cdd8e in m0_rm_remote_fini (rem=rem@entry=0x7f8ac00f5640) at rm/rm.c:1435
#8  0x00007f8aca3cde8e in remote_free (rem=0x7f8ac00f5640) at rm/rm.c:1377
#9  0x00007f8aca3cdac4 in m0_rm_loan_fini (loan=loan@entry=0x7f8aa808a768) at rm/rm.c:1264
#10 0x00007f8aca3cdb2e in m0_rm_outgoing_fini (out=out@entry=0x7f8aa808a760) at rm/rm.c:1169
#11 0x00007f8aca3cf423 in owner_balance (o=o@entry=0x7f8aa80871b0) at rm/rm.c:1951
#12 0x00007f8aca3d1220 in m0_rm_outgoing_complete (og=og@entry=0x7f8aa808a760) at rm/rm.c:2446
#13 0x00007f8acba40d47 in outreq_fini (rc=<optimized out>, outreq=0x7f8aa808a760) at /data/motr/rm/rm_fops.c:269
#14 rm_revoke_ast (grp=0x7f8aa8086ce0, ast=0x7f8aa808a800) at /data/motr/rm/rm_fops.c:494
#15 0x00007f8aca3f12c2 in m0_sm_asts_run (grp=grp@entry=0x7f8aa8086ce0) at sm/sm.c:169
#16 0x00007f8aca3f133d in m0_sm_group_lock (grp=grp@entry=0x7f8aa8086ce0) at sm/sm.c:86
#17 0x00007f8aca3d26f4 in credit_processor (rt=0x7f8aa8086c70) at rm/rm.c:702
#18 0x00007f8aca35df8e in m0_thread_trampoline (arg=arg@entry=0x7f8aa8086db8) at lib/thread.c:117
#19 0x00007f8aca368abd in uthread_trampoline (arg=0x7f8aa8086db8) at lib/user_space/uthread.c:98
#20 0x00007f8ac9ae7ea5 in start_thread () from /lib64/libpthread.so.0
#21 0x00007f8ac8bd896d in clone () from /lib64/libc.so.6
(gdb) 
```

Reply:
Another panic from m0d process:

```
#3  0x00007ff94f75cee4 in m0_panic (ctx=ctx@entry=0x7ff94fbf5c00 <__pctx.5769>) at lib/assert.c:52
#4  0x00007ff94f75e6ea in m0_clink_cleanup_locked (link=link@entry=0x7ff9440b5020) at lib/chan.c:321
#5  0x00007ff94f7dc350 in m0_rm_ha_unsubscribe (tracker=tracker@entry=0x7ff9440b4fc8) at rm/rm_ha.c:420
#6  0x00007ff94f7dc382 in m0_rm_ha_unsubscribe_lock (tracker=tracker@entry=0x7ff9440b4fc8) at rm/rm_ha.c:441
#7  0x00007ff94f7d3d8e in m0_rm_remote_fini (rem=rem@entry=0x7ff9440b4ee0) at rm/rm.c:1435
#8  0x00007ff94f7d3e8e in remote_free (rem=0x7ff9440b4ee0) at rm/rm.c:1377
#9  0x00007ff94f7d3ac4 in m0_rm_loan_fini (loan=loan@entry=0x7ff9440b5d20) at rm/rm.c:1264
#10 0x00007ff94f7d7777 in m0_rm_owner_loan_debit (owner=owner@entry=0x7ff9440b2f50, paid_loan=paid_loan@entry=0x7ff9440b4da0, 
    list=list@entry=0x7ff9440b3000) at rm/rm.c:2706
#11 0x00007ff94f7d8497 in m0_rm_loan_settle (owner=owner@entry=0x7ff9440b2f50, loan=loan@entry=0x7ff9440b4da0) at rm/rm.c:2804
#12 0x00007ff94f7d9a8e in cancel_process (fom=0x7ff9240536d0) at rm/rm_foms.c:778
#13 request_fom_tick (fom=0x7ff9240536d0, type=<optimized out>) at rm/rm_foms.c:667
#14 0x00007ff94f73205b in fom_exec (fom=0x7ff9240536d0) at fop/fom.c:785
#15 loc_handler_thread (th=0x2213cc0) at fop/fom.c:925
```

Reply:
If we do only `m0_clink_cleanup()`:

```diff
@@ -417,10 +417,7 @@ M0_INTERNAL void m0_rm_ha_unsubscribe(struct m0_rm_ha_tracker *tracker)
        struct m0_conf_obj *obj;
 
        if (m0_clink_is_armed(&tracker->rht_conf_exp)) {
-               if (m0_chan_is_locked(tracker->rht_conf_exp.cl_chan))
-                       m0_clink_cleanup_locked(&tracker->rht_conf_exp);
-               else
-                       m0_clink_cleanup(&tracker->rht_conf_exp);
+               m0_clink_cleanup(&tracker->rht_conf_exp);
        }
        if (m0_clink_is_armed(&tracker->rht_clink)) {
                obj = M0_AMB(obj, tracker->rht_clink.cl_chan, co_ha_chan);
```

Another UT panics:

```
$ ./scripts/m0 run-ut -t rm-rcredits-ut
----- run_ut -t rm-rcredits-ut -----
rm-rcredits-ut 
  no-borrow-flag                                  0.76 sec  248 MiB
  simple-borrow                                   0.71 sec  231 MiB
  two-borrows-single-req                          0.72 sec  232 MiB
  borrow-revoke-single-req                        0.73 sec  232 MiB
  revoke-with-hop                                 0.70 sec  231 MiB
  revoke-conflicting-wait                         0.70 sec  231 MiB
  revoke-conflicting-try                          0.69 sec  231 MiB
  borrow-non-conflicting                          0.69 sec  231 MiB
  revoke-no-conflict-wait                         0.69 sec  231 MiB
  revoke-no-conflict-try                          0.67 sec  232 MiB
  borrow-held-no-conflict                         0.66 sec  232 MiB
  borrow-held-conflicting                         0.70 sec  232 MiB
  starvation                                      0.74 sec  232 MiB
  barrier                                         1.67 sec  232 MiB
  barrier-overcome-star                           0.70 sec  232 MiB
  barrier-overcome-chain                          0.69 sec  232 MiB
  barrier-same-time                               0.67 sec  232 MiB
  debtor-death                                    0.67 sec  231 MiB
  creditor-death  motr[05693]:  1bf0  FATAL  [lib/assert.c:50:m0_panic]  panic: (m0_mutex_is_not_locked(mutex)) at m0_mutex_lock() (lib/mutex.c:53)  [git: sage-base-1.0-172-gc0812e366] /var/motr/m0ut/m0trace.5693
Motr panic: (m0_mutex_is_not_locked(mutex)) at m0_mutex_lock() lib/mutex.c:53 (errno: 2) (last failed: none) [git: sage-base-1.0-172-gc0812e366] pid: 5693  /var/motr/m0ut/m0trace.5693
...
$ sudo gdb ut/.libs/lt-m0ut /var/crash/core.0.lt-m0ut.5693
...
(gdb) bt
#0  0x00007f510b9b1387 in raise () from /lib64/libc.so.6
#1  0x00007f510b9b2a78 in abort () from /lib64/libc.so.6
#2  0x00007f510d208e25 in m0_arch_panic (c=c@entry=0x7f510d6939c0 <__pctx.4982>, ap=ap@entry=0x7ffe75461c68)
    at lib/user_space/uassert.c:131
#3  0x00007f510d1f7ee4 in m0_panic (ctx=ctx@entry=0x7f510d6939c0 <__pctx.4982>) at lib/assert.c:52
#4  0x00007f510d1fe5f4 in m0_mutex_lock (mutex=0x7f5113313fb8 <rm_ctxs+79032>) at lib/mutex.c:53
#5  0x00007f510d1f8fbd in m0_chan_lock (ch=<optimized out>) at lib/chan.c:70
#6  0x00007f510d1f9709 in m0_clink_cleanup (link=link@entry=0x17b3640) at lib/chan.c:313
#7  0x00007f510d277350 in m0_rm_ha_unsubscribe (tracker=tracker@entry=0x17b35e8) at rm/rm_ha.c:420
#8  0x00007f510d277382 in m0_rm_ha_unsubscribe_lock (tracker=tracker@entry=0x17b35e8) at rm/rm_ha.c:441
#9  0x00007f510d26ed8e in m0_rm_remote_fini (rem=0x17b3500) at rm/rm.c:1435
#10 0x00007f510e8e4af5 in rm_ctx_server_windup (srv_id=srv_id@entry=SERVER_2) at rm/ut/rmut.c:287
#11 0x00007f510e8e4b99 in rm_ctx_server_owner_windup (srv_id=srv_id@entry=SERVER_2) at rm/ut/rmut.c:275
#12 0x00007f510e8dd61f in remote_credits_utfini () at rm/ut/rcredits.c:206
#13 0x00007f510e91ab9e in run_test (max_name_len=43, test=0x696368 <rm_rcredits_ut+5064>) at ut/ut.c:389
#14 run_suite (max_name_len=43, suite=0x694fa0 <rm_rcredits_ut>) at ut/ut.c:454
#15 tests_run_all (m=0x7f5118f55a80 <ut.12177>) at ut/ut.c:507
#16 m0_ut_run () at ut/ut.c:533
#17 0x0000000000404436 in main (argc=3, argv=0x7ffe75462578) at ut/m0ut.c:502
(gdb) 
```

Reply:
@nikitadanilov, please review https://github.com/Seagate/cortx-motr/pull/304/commits/c87a2c396760bee4336419caf691c5e813a588b6 for this.

Reply:
I'm only not sure whether it will work in case of the actual configuration expiration path, probably not:

```
m0_confc_expired_cb() ->
  m0_chan_broadcast_lock(&reqh->rh_conf_cache_exp) ->
    rm_ha_conf_expired_cb() ->
      m0_rm_ha_unsubscribe_lock() -> ...
```

Reply:
@nikitadanilov, please review the updated patch for this at https://github.com/Seagate/cortx-motr/pull/304/commits/01162dad04f208ce81572a36acdecacdfae00c32.